File: runtime-python/helix_rt/batching.py
import asyncio
from typing import List, Any, Callable, Awaitable, Optional

class BatchRequest:
    def __init__(self, req: Any):
        self.req = req
        self.future = asyncio.get_event_loop().create_future()

class BatchScheduler:
    """
    Helix RPC Dynamic Batch Scheduler for Python asyncio.
    Absorbs concurrent requests into a unified List, dispatching them either
    when max_size is hit or the batch_window_ms timer expires.
    """
    def __init__(self, max_size: int, batch_window_ms: float, handler: Callable[[List[Any]], Awaitable[List[Any]]]):
        self.max_size = max_size
        self.batch_window_ms = batch_window_ms / 1000.0
        self.handler = handler
        
        self.queue: List[BatchRequest] = []
        self.lock = asyncio.Lock()
        self.timer_task: Optional[asyncio.Task] = None

    async def invoke(self, req: Any) -> Any:
        br = BatchRequest(req)
        
        async with self.lock:
            self.queue.append(br)
            
            if len(self.queue) >= self.max_size:
                # Dispatch immediately
                dispatch_queue = self.queue
                self.queue = []
                if self.timer_task and not self.timer_task.done():
                    self.timer_task.cancel()
                asyncio.create_task(self._process_batch(dispatch_queue))
            elif len(self.queue) == 1:
                # First item, start timer
                loop = asyncio.get_running_loop()
                self.timer_task = loop.create_task(self._wait_and_dispatch())
                
        # Wait for the batch to be processed and return the specific result
        return await br.future
        
    async def _wait_and_dispatch(self):
        try:
            await asyncio.sleep(self.batch_window_ms)
        except asyncio.CancelledError:
            return
            
        async with self.lock:
            if not self.queue:
                return
            dispatch_queue = self.queue
            self.queue = []
            
        await self._process_batch(dispatch_queue)
        
    async def _process_batch(self, batch: List[BatchRequest]):
        reqs = [b.req for b in batch]
        try:
            resps = await self.handler(reqs)
            if len(resps) != len(batch):
                raise ValueError("Batch Handler returned mismatched response list length")
            for br, resp in zip(batch, resps):
                if not br.future.done():
                    br.future.set_result(resp)
        except Exception as e:
            for br in batch:
                if not br.future.done():
                    br.future.set_exception(e)

File: runtimes/rust/src/batching.rs
use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;
use tokio::sync::{oneshot, RwLock};

/// Dynamic batch request
struct BatchRequest {
    payload: Vec<u8>,
    is_json: bool,
    tx: oneshot::Sender<Result<(Vec<u8>, String), String>>,
}

/// BatchScheduler collects concurrent requests into a single dispatch batch.
/// When the batch window expires OR max_size is reached, all pending requests
/// are dispatched together to the handler and results are fanned back out.
pub struct BatchScheduler {
    tx: tokio::sync::mpsc::Sender<BatchRequest>,
}

impl BatchScheduler {
    pub fn new<H>(max_size: usize, window_ms: u64, handler: H) -> Self
    where
        H: Fn(Vec<Vec<u8>>, bool) -> Result<Vec<Vec<u8>>, String> + Send + Sync + 'static,
    {
        let (tx, mut rx) = tokio::sync::mpsc::channel::<BatchRequest>(max_size * 100);
        let handler = Arc::new(handler);

        tokio::spawn(async move {
            let mut batch: Vec<BatchRequest> = Vec::new();

            loop {
                // Wait for the first item
                match rx.recv().await {
                    None => break,
                    Some(req) => {
                        let is_json = req.is_json;
                        batch.push(req);

                        // Drain more items up to max_size within the window
                        let window = tokio::time::Duration::from_millis(window_ms);
                        let deadline = tokio::time::Instant::now() + window;

                        loop {
                            if batch.len() >= max_size {
                                break;
                            }
                            match tokio::time::timeout_at(deadline, rx.recv()).await {
                                Ok(Some(r)) => batch.push(r),
                                _ => break,
                            }
                        }

                        // Dispatch the batch
                        let payloads: Vec<Vec<u8>> =
                            batch.iter().map(|r| r.payload.clone()).collect();
                        let h = handler.clone();
                        let results =
                            tokio::task::spawn_blocking(move || h(payloads, is_json)).await;

                        match results {
                            Ok(Ok(resps)) if resps.len() == batch.len() => {
                                for (req, resp) in batch.drain(..).zip(resps) {
                                    let _ = req.tx.send(Ok((
                                        resp,
                                        if is_json {
                                            "application/json".to_string()
                                        } else {
                                            "application/grpc".to_string()
                                        },
                                    )));
                                }
                            }
                            Ok(Ok(_)) => {
                                for req in batch.drain(..) {
                                    let _ = req.tx.send(Err(
                                        "batch handler returned mismatched response count"
                                            .to_string(),
                                    ));
                                }
                            }
                            Ok(Err(e)) => {
                                for req in batch.drain(..) {
                                    let _ = req.tx.send(Err(e.clone()));
                                }
                            }
                            Err(e) => {
                                let msg = e.to_string();
                                for req in batch.drain(..) {
                                    let _ = req.tx.send(Err(msg.clone()));
                                }
                            }
                        }
                    }
                }
            }
        });

        BatchScheduler { tx }
    }

    pub async fn invoke(
        &self,
        payload: Vec<u8>,
        is_json: bool,
    ) -> Result<(Vec<u8>, String), String> {
        let (tx, rx) = oneshot::channel();
        self.tx
            .send(BatchRequest {
                payload,
                is_json,
                tx,
            })
            .await
            .map_err(|e| e.to_string())?;
        rx.await.map_err(|e| e.to_string())?
    }
}

/// LeastConnBalancer is a Rust port of the lock-free Go implementation.
/// It uses a RwLock-protected pre-indexed map with AtomicI64 counters per target.
pub struct LeastConnBalancer {
    endpoints: Arc<RwLock<HashMap<String, Arc<AtomicI64>>>>,
}

impl LeastConnBalancer {
    pub fn new() -> Self {
        LeastConnBalancer {
            endpoints: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn register(&self, targets: &[&str]) {
        let mut map = self.endpoints.write().await;
        for t in targets {
            map.entry(t.to_string())
                .or_insert_with(|| Arc::new(AtomicI64::new(0)));
        }
    }

    pub async fn next(&self, targets: &[String]) -> Result<String, String> {
        if targets.is_empty() {
            return Err("no targets available".to_string());
        }

        let map = self.endpoints.read().await;

        // Ensure all targets registered (register new ones via write lock if needed)
        let mut selected = targets[0].clone();
        let mut min_active = i64::MAX;

        for t in targets {
            let counter = map.get(t.as_str());
            let active = counter.map(|c| c.load(Ordering::Relaxed)).unwrap_or(0);
            if active < min_active {
                min_active = active;
                selected = t.clone();
            }
        }

        // Increment winner
        if let Some(counter) = map.get(selected.as_str()) {
            counter.fetch_add(1, Ordering::Relaxed);
        }

        Ok(selected)
    }

    pub async fn done(&self, target: &str) {
        let map = self.endpoints.read().await;
        if let Some(counter) = map.get(target) {
            counter.fetch_sub(1, Ordering::Relaxed);
        }
    }
}

impl Default for LeastConnBalancer {
    fn default() -> Self {
        Self::new()
    }
}
